torch/lib/THD/base/data_channels/GlooCache.hpp

```diff
@@ -58,7 +58,8 @@ struct GlooCache {
   using value_type = std::tuple<
     std::shared_ptr<algorithm_type>, // algorithm
     std::shared_ptr<buffer_type>,    // input buffer (nullptr if not used)
-    std::shared_ptr<buffer_type>     // output buffer (nullptr if not used)
+    std::shared_ptr<buffer_type>,    // output buffer (nullptr if not used)
+    std::shared_ptr<std::mutex>      // mutex to make algorithms run atomically
```

Comment:
Same here, it's just so you don't use the same algorithm concurrently, we don't care if the effects are seen atomically. We don't depend on that